### ISSUE TITLE
[Android] NetworkInterface.GetAllNetworkInterfaces failed with NetworkInformationException (message: Success)

### ISSUE THREAD

**Issue by 8212369** [dotnet/runtime#75809]
### Description

On Android 11 release build, I was encountering System.Net.NetworkInformation.NetworkInformationException when called System.Net.NetworkInformation.NetworkInterface。GetAllNetworkInterfaces, and the error string is just success. I have added two permissions: INTERNET and ACCESS_NETWORK_STATE to which no avail. I think it's worth mentioning.

I saw a log category named DOTNET_LINK so I reported the issue here.

### Reproduction Steps

Using System.Net namespace, call NetworkInterface.GetAllNetworkInterfaces

### Expected behavior

The function completes with no exception

### Actual behavior

The function throws System.Net.NetworkInformationException 

![image](https://user-images.githubusercontent.com/99397963/190917233-7e2ca66d-d8d6-4f43-9d2b-d7ee5e31d3c1.png)

### Configuration

.NET 6.0.9

### Other information

I skimmed through the PRs and there was a change to exclusively use the Xamarin's implementation of getifaddrs recently. But on Android 11 Google [blocked NETLINK_ROUTE](https://developer.android.com/training/articles/user-data-ids) family socket from being bind.

This is also the log that I was getting from adb

![image](https://user-images.githubusercontent.com/99397963/190917055-c92d23ea-c266-4c47-9cb2-faf17a901807.png)

I think it's correspond to [this](https://github.com/dotnet/runtime/blob/main/src/native/libs/System.Native/pal_ifaddrs.c#L103). I hope I'm right.

One small note, I think this quirk is per vendor distribution, as my Xiaomi phone does not seems have this problem.

Related [issues](https://github.com/arvidn/libtorrent/issues/6251) about this blocking .

**Comment by danmoseley** [comment]
Thanks - could you paste in the stacks/messages as regular text? You can use triple back tick above and below for formatting. This makes it possible for search to find them

**Comment by 8212369** [comment]
Here is the exception stack:

```
[ERROR] FATAL UNHANDLED EXCEPTION: System.Net.NetworkInformation.NetworkInformationException (0x80004005): Success
   at System.Net.NetworkInformation.NetworkInterfacePal.TransformNetworkInterfacess[AndroidNetworkInterface[]](Func`5 transform)
   at System.Net.NetworkInformation.NetworkInterfacePal.GetAndroidNetworkInterfaces()
   at System.Net.NetworkInformation.NetworkInterfacePal.GetAllNetworkInterfaces()
   at System.Net.NetworkInformation.NetworkInterface.GetAllNetworkInterfaces()
   at Microsoft.Phone.Net.NetworkInformation.NetworkInterface.get_NetworkInterfaceType()
   at TestApp.MetricsSender.CheckConnectionType()
   at System.Threading.Thread.StartCallback()
```

The log message I think to be related is this line. This is the only error line I found near this exception:

https://github.com/dotnet/runtime/blob/55060ebf68e16f2838d30060d41cd71b0bc9c540/src/native/libs/System.Native/pal_ifaddrs.c#L104

**Comment by forlayo** [comment]
Another one suffering the issue here , this is what I see on logcat..

```
...
09-20 20:38:26.869  6067  6110 W DOTNET_NETLINK: Failed to bind to the netlink socket. Permission denied
...
09-20 20:38:26.865  6067  6067 W com.displaynote.amadeus: type=1400 audit(0.0:23150): avc: denied { bind } for comm=2E4E455420546872656164506F6F6C scontext=u:r:untrusted_app:s0:c168,c256,c512,c768 tcontext=u:r:untrusted_app:s0:c168,c256,c512,c768 tclass=netlink_route_socket permissive=0 b/155595000 app=com.displaynote.amadeus
09-20 20:38:26.889  6067  6110 E System.Net.NetworkInformation.NetworkInformationException (0x80004005): Success
09-20 20:38:26.889  6067  6110 E Amadeus.Services.Discovery.Method.CloudDiscovery:    at System.Net.NetworkInformation.NetworkInterfacePal.TransformNetworkInterfacess[AndroidNetworkInterface[]](Func`5 transform)
09-20 20:38:26.889  6067  6110 E Amadeus.Services.Discovery.Method.CloudDiscovery:    at System.Net.NetworkInformation.NetworkInterfacePal.GetAndroidNetworkInterfaces()
09-20 20:38:26.889  6067  6110 E Amadeus.Services.Discovery.Method.CloudDiscovery:    at System.Net.NetworkInformation.NetworkInterfacePal.GetAllNetworkInterfaces()
09-20 20:38:26.889  6067  6110 E Amadeus.Services.Discovery.Method.CloudDiscovery:    at System.Net.NetworkInformation.NetworkInterface.GetAllNetworkInterfaces()
```

**Comment by simonrozsival** [comment]
I started looking into the problem and as a quick workaround it is possible to modify `AndroidManifest.xml` with:
```xml
<uses-sdk android:minSdkVersion="29" />
```

When the `minSdkVersion` is 29 or lower, the exception isn't thrown and a correct array of interfaces is returned (tested on a Samsung phone with Android 12). In the case when it is necessary to use 30 or newer, it is necessary to use [`Java.Net.NetworkInterface.NetworkInterfaces`](https://learn.microsoft.com/en-us/dotnet/api/java.net.networkinterface.networkinterfaces?view=xamarin-android-sdk-12).

I'll keep working on a proper fix for this issue that will work with new SDKs.

**Comment by forlayo** [comment]
@simonrozsival If you are looking for a workaround, without changing minSdkVersion you can do something like this, which is what I am doing:

``` 
    public IList<IPAddress> GetAllLocalValidIp4Addresses()
    {
#if ANDROID

        // This is a workaround for this issue of MAUI on Android -> https://github.com/dotnet/runtime/issues/75809 
        // Instead of using regular .net we use Java Interop bindings directly.

        var inetEnum = Java.Net.NetworkInterface.NetworkInterfaces;
        if (inetEnum is null)
        {
            return new List<IPAddress>();
        }

        var ipAddresses = new List<IPAddress>();
        foreach (var interfaces in Java.Util.Collections.List(inetEnum))
        {
            var addresses = (interfaces as Java.Net.NetworkInterface)?.InetAddresses;
            if (addresses == null)
            {
                continue;
            }

            foreach (Java.Net.InetAddress address in Java.Util.Collections.List(addresses))
            {
                if (address.HostAddress == null || address.IsLoopbackAddress || address is not Java.Net.Inet4Address)
                {
                    continue;
                }
                ipAddresses.Add(IPAddress.Parse(address.HostAddress));
            }
        }

        return ipAddresses;
#else
#pragma warning disable CS8619 // This is not going to have any null, but the compiler is not able to determine it.
        return NetworkInterface
            .GetAllNetworkInterfaces()
            .Where(n => n.OperationalStatus == OperationalStatus.Up)
            .Where(n => n.NetworkInterfaceType == NetworkInterfaceType.Wireless80211 || n.NetworkInterfaceType == NetworkInterfaceType.Ethernet)
            .Where(n => !n.Description.ToLower().Contains("virtual"))
            .Where(n => !n.Description.ToLower().Contains("pseudo"))
            .SelectMany(n => n.GetIPProperties().UnicastAddresses)
            .Where(n => n.Address.AddressFamily == AddressFamily.InterNetwork)
            .Select(g => g?.Address)
            .Where(a => a != null) // Here I am checking to not adding anything null.
            .ToList();
#pragma warning restore CS8619
#endif
    }
```

**Comment by Slm0nB** [comment]
We are struggling with this issue as well, and we looking forward to a fix.  Thank you for the work around - appreciated!

**Comment by jahmai-ca** [comment]
@simonrozsival I can see this has been back-ported to net7, but it also affects net6, and net7 isn't released yet. Can we get a back port to net6 please?

**Comment by MartinRothschink** [comment]
I second that, we **need net6 support**.
That function was always not working correctly on Android: https://github.com/dotnet/maui/issues/6649

**Comment by steveisok** [comment]
This was an oversight on my part as I only backported our change to .NET 7.  I'll work to get this back to .NET 6.

**Comment by karelz** [comment]
Fixed in main (8.0) in PRs #76370 and #76541, and in 7.0 (RC2?) in PR #76565 and in 6.0.12 in PR #77260.
